File: JavaScript/239-Sliding Window Maximum.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var maxSlidingWindow = function(nums, k) {
    let res = [];
    let stack = [];
    for (let i=0; i<nums.length; i++){
        while (stack.length && nums[i] > nums[stack[stack.length-1]]){
            stack.pop();
        }
        stack.push(i)
        if (i < k-1){
            continue;
        }
        if (i - stack[0] === k){
            stack.shift();
        }
        res.push(nums[stack[0]]);
    }
    return res;
};

File: 239-Sliding Window Maximum.py
from collections import deque
class Solution(object):
    def maxSlidingWindow(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        stack = deque()
        res = []
        
        for i in range(len(nums)):
            while stack and nums[i] > nums[stack[-1]]:
                stack.pop()
            stack.append(i)
            if i < k-1:
                continue
            if i - stack[0] == k:
                stack.popleft()
            res.append(nums[stack[0]])
        
        return res